File: app/src/main/java/com/mario/template/ui/layout/view/calendar/BaseCalendarView.java
package com.mario.template.ui.layout.view.calendar;

import android.content.Context;
import android.content.res.Resources;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.mario.template.R;

import org.jetbrains.annotations.NotNull;

import java.util.Calendar;
import java.util.Date;

public class BaseCalendarView extends FrameLayout {
    public final static int WEEK_VIEW_H = 50;

    private LinearLayout weeksLayout;
    private RecyclerView daysRecyclerView;
    private MonthCalendarAdapter adapter;
    private MonthCalendarAdapter.Listener listener = null;

    public BaseCalendarView(Context context) {
        super(context);
    }

    public BaseCalendarView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0, 0);
    }

    public BaseCalendarView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(
                context,
                attrs,
                defStyleAttr
        );
        init(attrs, defStyleAttr, defStyleAttr);
    }

    private void init(AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        initWeekView();
        initDayView();
    }

    public static int dpToPx(Context context, int dp) {
        Resources resources = context.getResources();

        DisplayMetrics metrics = resources.getDisplayMetrics();

        float px = dp * (metrics.densityDpi / 160f);

        return (int) px;
    }

    private void initWeekView() {
        weeksLayout = new LinearLayout(getContext());

        String[] arrWeekName = getResources().getStringArray(R.array.week);
        for (int i = 0; i < BaseCalendar.DAYS_OF_WEEK; i++) {
            TextView weekView = new TextView(getContext());
            weekView.setText(arrWeekName[i]);

            if (i == 0) {
                weekView.setTextColor(getResources().getColor(R.color.colorAccent));
            } else {
                weekView.setTextColor(getResources().getColor(R.color.colorAccent));
            }
            weekView.setTextSize(16.0f);
            weekView.setGravity(Gravity.CENTER);

            LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
            layoutParams.weight = 1.0f;
            weeksLayout.addView(weekView, layoutParams);
        }

        LayoutParams layoutParams = new LayoutParams(LayoutParams.MATCH_PARENT, dpToPx(getContext(), WEEK_VIEW_H));
        addView(weeksLayout, layoutParams);
    }


    private void initDayView() {
        daysRecyclerView = new RecyclerView(getContext());
        daysRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), BaseCalendar.DAYS_OF_WEEK));

        adapter = new MonthCalendarAdapter(getContext(), new MonthCalendarAdapter.Listener() {

            @Override
            public void refreshCurrentMonth(@NotNull Calendar calendar) {
                listener.refreshCurrentMonth(calendar);
            }

            @Override
            public void selecteDate(@NotNull Date date) {
                listener.selecteDate(date);
            }
        });
        daysRecyclerView.setAdapter(adapter);
        adapter.notifyDataSetChanged();

        LayoutParams layoutParams = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
        layoutParams.topMargin = dpToPx(getContext(), WEEK_VIEW_H);
        addView(daysRecyclerView, layoutParams);
    }

    public void setListener(MonthCalendarAdapter.Listener listener) {
        this.listener = listener;
    }

    public void nextMonth() {
        adapter.changeToNextMonth();
    }

    public void prevMonth() {
        adapter.changeToPrevMonth();
    }
}

File: app/src/main/java/com/victoria/bleled/app/main/calendar/BaseCalendarView.kt
package com.victoria.bleled.app.main.calendar

import android.content.Context
import android.util.AttributeSet
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.victoria.bleled.R
import com.victoria.bleled.util.CommonUtil
import java.util.*

class BaseCalendarView : FrameLayout {

    companion object {
        const val WEEK_VIEW_H = 30
    }

    private lateinit var weeksLayout: LinearLayout
    private lateinit var daysRecyclerView: RecyclerView
    private lateinit var adapter: MonthCalendarAdapter


    constructor(context: Context) : super(context) {

    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(attrs, defStyleAttr, defStyleAttr)
    }

    private fun init(attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) {
        initWeekView()
        initDayView()
    }

    private fun initWeekView() {
        weeksLayout = LinearLayout(context)

        val arrWeekName = resources.getStringArray(R.array.week)
        for (i in 0 until BaseCalendar.DAYS_OF_WEEK) {
            var weekView = TextView(context)
            weekView.text = arrWeekName[i]

            if (i == 0) {
                weekView.setTextColor(resources.getColor(R.color.colorAccent))
            } else {
                weekView.setTextColor(resources.getColor(R.color.color_text))
            }
            weekView.textSize = 20f

            val layoutParams =
                LinearLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
            layoutParams.weight = 1f
            weeksLayout.addView(weekView, layoutParams)
        }

        val layoutParams =
            LayoutParams(LayoutParams.MATCH_PARENT, CommonUtil.dpToPx(context, WEEK_VIEW_H))
        addView(weeksLayout, layoutParams)
    }

    private fun initDayView() {
        daysRecyclerView = RecyclerView(context)
        daysRecyclerView.layoutManager =
            GridLayoutManager(context, BaseCalendar.DAYS_OF_WEEK)

        adapter = MonthCalendarAdapter(context, object : MonthCalendarAdapterListener {
            override fun refreshCurrentMonth(calendar: Calendar) {
                //val sdf = SimpleDateFormat("yyyy MM", Locale.KOREAN)
                //viewDataBinding.tvMonth.text = sdf.format(calendar.time)
            }
        })
        daysRecyclerView.adapter = adapter
        adapter.notifyDataSetChanged()

        val layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
        layoutParams.topMargin = CommonUtil.dpToPx(context!!, WEEK_VIEW_H)
        addView(daysRecyclerView, layoutParams)
    }
}